#1130: Diagnostics/debug info for AsyncHttpClient

Diagoras opened the pull request:
This is a first crack at addressing that issue, with only information on the total number of connections and how many are active or idle. At the very least, it'd be nice to also see the same data on a per host basis, but I thought I'd get what I've done so far looked at first.

Diagoras commented on client/src/test/java/org/asynchttpclient/ClientStatsTest.java:
```diff
@@ -0,0 +1,56 @@
+
+            for (ListenableFuture<Response> future : futures) {
+                future.get();
+            }
```
I'd very much like to add an assert here when the connections are active, but the request/response cycle happens so fast I haven't been able to pull the ClientStats when the requests are still in flight. Is there a way to configure the server that's created by this test to delay responses?

slandelle replied:
Have a look at `EchoHandler`, which is the default handler used in the Jetty servers.
It can use a `LockThread` header to block the response on the server side. It current performs a hard coded 40s sleep, but it's actually no longer used. It could be used to pass the sleep value instead.

Diagoras commented on client/src/test/java/org/asynchttpclient/ClientStatsTest.java:
```diff
@@ -0,0 +1,56 @@
+            assertEquals(idleStats.getTotalConnectionCount(), 5);
+            assertEquals(idleStats.toString(), "There are 5 total connections, 0 are active and 5 are idle.");
+
+            Thread.sleep(5000);
```
Bleh. :(

slandelle replied:
I don't think there's any better solution, as you want to wait for the pool idle timeout to kick in.

slandelle commented on client/src/test/java/org/asynchttpclient/ClientStatsTest.java:
```diff
@@ -0,0 +1,56 @@
+package org.asynchttpclient;
```
missing file header

slandelle replied:
still missing

slandelle commented on client/src/main/java/org/asynchttpclient/ClientStats.java:
```diff
@@ -0,0 +1,37 @@
+package org.asynchttpclient;
```
missing file header

slandelle commented:
Thanks!
Please address comments and ping once you're done.

slandelle commented on client/src/main/java/org/asynchttpclient/ClientStats.java:
```diff
@@ -0,0 +1,37 @@
+
+/**
+ * Created by grenville on 9/24/16.
+ */
```
please remove

slandelle commented:
@Diagoras gentle ping

Diagoras commented:
@slandelle Sorry, things have been a bit crazy IRL. Should push a commit addressing your comments sometime this weekend.

slandelle commented on client/src/main/java/org/asynchttpclient/DefaultAsyncHttpClient.java:
```diff
@@ -255,6 +257,13 @@ public EventLoopGroup getEventLoopGroup() {
     }
 
+    public ClientStats getClientStats() {
+        final ChannelManager channelManager = requestSender.getChannelManager();
```
You don't need that, `DefaultAsyncHttpClient` has the `ChannelManager`.

slandelle commented on client/src/main/java/org/asynchttpclient/netty/request/NettyRequestSender.java:
```diff
@@ -536,4 +536,8 @@ public void call() {
         Channels.setAttribute(channel, newExecuteNextRequestCallback(future, nextRequest));
     }
+
+    public ChannelManager getChannelManager() {
```
Useless, `DefaultAsyncHttpClient` has the `ChannelManager`.

slandelle commented on client/src/main/java/org/asynchttpclient/netty/channel/ChannelManager.java:
```diff
@@ -491,4 +491,8 @@ public ChannelPool getChannelPool() {
         return eventLoopGroup;
     }
+
+    public ChannelGroup getOpenChannels() {
```
No, this shouldn't leak

slandelle commented on client/src/main/java/org/asynchttpclient/DefaultAsyncHttpClient.java:
```diff
@@ -255,6 +257,13 @@ public EventLoopGroup getEventLoopGroup() {
         return channelManager.getEventLoopGroup();
     }
 
+    public ClientStats getClientStats() {
```
Performing the computation should be the responsibility of `ChannelManager`, not `DefaultAsyncHttpClient`.

slandelle commented on client/src/main/java/org/asynchttpclient/DefaultAsyncHttpClient.java:
```diff
@@ -255,6 +257,13 @@ public EventLoopGroup getEventLoopGroup() {
+    public ClientStats getClientStats() {
+        final ChannelManager channelManager = requestSender.getChannelManager();
+        final long activeConnectionCount = channelManager.getOpenChannels().stream().filter(channel -> !channel.isOpen()).count();
+        final long idleConnectionCount = channelManager.getOpenChannels().stream().filter(Channel::isOpen).count();
```
It should be possible to only scan once, not twice.

slandelle replied:
I really wonder if it's possible to get `idleConnectionCount` different from zero.

slandelle commented on client/src/main/java/org/asynchttpclient/ClientStats.java:
```diff
@@ -0,0 +1,37 @@
+        return "There are " + totalConnectionCount +
+                " total connections, " + activeConnectionCount +
+                " are active and " + idleConnectionCount + " are idle.";
+    }
```
implement `equals` and `hashcode`

slandelle commented:
No pro, thanks!

Diagoras commented:
Hey, thanks for all the comments! I caught something nasty on Friday and have been out of commission this weekend. Hoping to address them on Monday/Tuesday.

slandelle commented:
No hurries

Diagoras commented:
So, Monday somehow became a month, but I guess better late than never. I believe I addressed all of your comments, but poke me if I missed one. I also have some questions/comments of my own which I'll add to this PR.

Diagoras commented on client/src/main/java/org/asynchttpclient/DefaultAsyncHttpClient.java:
```diff
@@ -33,6 +30,10 @@
+import io.netty.channel.EventLoopGroup;
+import io.netty.util.HashedWheelTimer;
+import io.netty.util.Timer;
+
```
...I guess my IDE is a little opinionated. Revert? Is there a style guide for this project, including a preferred import ordering?

Diagoras commented on client/src/main/java/org/asynchttpclient/netty/channel/ChannelManager.java:
```diff
@@ -488,4 +511,11 @@ public ChannelPool getChannelPool() {
     }
+
+    public ClientStats getClientStats() {
+        final long totalConnectionCount = openChannels.size();
```
Do we need to filter both this count and and the idle channel count below to ensure we don't count instances of `ServerChannel`?

Given that the `ChannelManager` can be set in the config, I imagine someone using Netty as their server *could* pass their `ChannelManager` in here. `openChannels` also explicitly contains two `Set`s, one of which is for server channels.

slandelle replied:
> Given that the ChannelManager can be set in the config

No, it can't.

Diagoras replied:
Hmm, okay. So I guess there's no need to filter out `ServerChannel` instances. Good to know.

Diagoras commented on client/src/main/java/org/asynchttpclient/netty/channel/DefaultChannelPool.java:
```diff
@@ -356,6 +357,16 @@ public void flushPartitions(ChannelPoolPartitionSelector selector) {
+    @Override
+    public long getIdleChannelCount() {
+        return partitions.reduceValuesToLong(
+                Long.MAX_VALUE,
```
Since this parameter decides how many elements are needed to activate parallelism, `Long.MAX_VALUE` pretty much switches off parallel reduction of the elements. Do you think that's too conservative?

Diagoras commented on client/src/main/java/org/asynchttpclient/netty/channel/DefaultChannelPool.java:
```diff
@@ -356,6 +357,16 @@ public void flushPartitions(ChannelPoolPartitionSelector selector) {
 
+    @Override
+    public long getIdleChannelCount() {
+        return partitions.reduceValuesToLong(
```
This method is neat, but examining the implementation shows that it involves blocking on a `ForkJoinTask`. Unless there's a particular reason to want it, it might be better to just call `partitions.values().stream().map(ConcurrentLinkedDeque::size).count()`.

slandelle replied:
Yeah, I'd go with summing `ConcurrentLinkedDeque::size`. Those stats would just be an estimate anyway, you can't get a proper snapshot without freezing, which we of course don't want.

slandelle commented on client/src/main/java/org/asynchttpclient/ClientStats.java:
```diff
@@ -0,0 +1,78 @@
+/*
+ * Copyright 2010 Ning, Inc.
```
Nope. Ning hasn't been involved in here for years. I've never even been in touch with them.
Code that goes back to their sponsorship still have their copyright, but all new code goes under `AsyncHttpClient Project`

slandelle commented on client/src/main/java/org/asynchttpclient/DefaultAsyncHttpClient.java:
```diff
@@ -255,6 +255,11 @@ public EventLoopGroup getEventLoopGroup() {
         return channelManager.getEventLoopGroup();
     }
 
+    @Override
```
Let's aim for AHC 2.1 and add it to the API

slandelle commented on client/src/main/java/org/asynchttpclient/ClientStats.java:
```diff
@@ -0,0 +1,78 @@
+    private final long idleConnectionCount;
+
+    public ClientStats(final long activeConnectionCount,
+                       final long idleConnectionCount) {
```
don't really need final here

Diagoras replied:
I always try to make my references immutable by default, but I can change it.

slandelle replied:
Just a small nit. But considering the code below, the finals are pretty useless here. God, Scala gets it right here (immutability is the default)...

Diagoras replied:
You can say that again.

slandelle commented on client/src/test/java/org/asynchttpclient/ClientStatsTest.java:
```diff
@@ -0,0 +1,169 @@
+
+    @Test(groups = "standalone")
+    public void testClientStatus() throws Throwable {
+        try (final DefaultAsyncHttpClient client = (DefaultAsyncHttpClient) asyncHttpClient(config().setKeepAlive(true).setPooledConnectionIdleTimeout(5000))) {
```
no need to cast if we aim for AHC 2.1 and change the `AsyncHttpClient` interface.

slandelle commented on client/src/test/java/org/asynchttpclient/ClientStatsTest.java:
```diff
@@ -0,0 +1,169 @@
+            assertEquals(emptyStats.getIdleConnectionCount(), 0);
+            assertEquals(emptyStats.getTotalConnectionCount(), 0);
+
+            final List<ListenableFuture<Response>> futures = new ArrayList<>();
```
I'll be honest: I rarely use Java 8 streams as I mostly use Scala, but isn't there a way to have this more concise with Java 8?

slandelle replied:
> Hey, me too!

Sounds like a Play! user :)

slandelle commented on client/src/test/java/org/asynchttpclient/ClientStatsTest.java:
```diff
@@ -0,0 +1,169 @@
+            // Let's make sure the active count is correct when reusing cached connections.
+
+            final List<ListenableFuture<Response>> repeatedFutures = new ArrayList<>();
+            for (int i = 0; i < 3; i++) {
```
with streams?

slandelle commented on client/src/test/java/org/asynchttpclient/ClientStatsTest.java:
```diff
@@ -0,0 +1,169 @@
+
+    @Test(groups = "standalone")
+    public void testClientStatusNoKeepalive() throws Throwable {
+        try (final DefaultAsyncHttpClient client = (DefaultAsyncHttpClient) asyncHttpClient(config().setKeepAlive(false))) {
```
don't cast

slandelle commented on extras/registry/src/test/java/org/asynchttpclient/extras/registry/BadAsyncHttpClient.java:
```diff
@@ -125,4 +126,9 @@ public BoundRequestBuilder prepareRequest(RequestBuilder requestBuilder) {
+
+    @Override
+    public ClientStats getClientStats() {
+        return null;
```
throw new UnsupportedOperationException

slandelle commented on extras/registry/src/test/java/org/asynchttpclient/extras/registry/TestAsyncHttpClient.java:
```diff
@@ -122,4 +123,8 @@ public BoundRequestBuilder prepareRequest(RequestBuilder requestBuilder) {
 
+    @Override
+    public ClientStats getClientStats() {
+        return null;
```
throw new UnsupportedOperationException

slandelle commented on client/src/test/java/org/asynchttpclient/ClientStatsTest.java:
```diff
@@ -0,0 +1,169 @@
+            assertEquals(emptyStats.getTotalConnectionCount(), 0);
+
+            final List<ListenableFuture<Response>> futures = new ArrayList<>();
+            for (int i = 0; i < 5; i++) {
```
stream

slandelle commented on client/src/test/java/org/asynchttpclient/ClientStatsTest.java:
```diff
@@ -0,0 +1,169 @@
+            // Let's make sure the active count is correct when reusing cached connections.
+
+            final List<ListenableFuture<Response>> repeatedFutures = new ArrayList<>();
+            for (int i = 0; i < 3; i++) {
```
stream

slandelle commented on client/src/main/java/org/asynchttpclient/channel/NoopChannelPool.java:
```diff
@@ -50,4 +50,11 @@ public void flushPartition(Object partitionKey) {
+        return 0;
+    }
+
+
```
kill empty lines

slandelle commented:
Hey! Thanks for updating! Sorry, I still have some comments...

slandelle commented on client/src/main/java/org/asynchttpclient/AsyncHttpClient.java:
```diff
@@ -266,4 +266,11 @@
+     *
+     * @return a {@link ClientStats}
+     */
+    ClientStats getClientStats();
```
Let's aim for AHC 2.1 so we can change this API.

Diagoras replied:
Then I should reopen this PR against branch "2.1", right?

slandelle replied:
No. The 2.1 branch will be dropped by a few weeks when I'll move the master to Netty 4.1. This branch was solely for the Play people so they can make progress on their own upgrade. It lags behind master in terms for bug fixes and refactoring.

Diagoras commented:
I added per host stats, so I think we're feature complete unless you can think of anything else people might want.

Diagoras commented on client/src/main/java/org/asynchttpclient/ClientStats.java:
```diff
@@ -0,0 +1,92 @@
+    private final Map<String, HostStats> statsPerHost;
+
+    public ClientStats(Map<String, HostStats> statsPerHost) {
+        this.statsPerHost = Collections.unmodifiableMap(statsPerHost);
```
Java doesn't have immutable collections and Guava (with its ImmutableMap) isn't in scope, so I went with this. Sadly, there's no way to have the type reflect that this Map is immutable, so it's noted in the Javadoc instead.

Diagoras commented on client/src/main/java/org/asynchttpclient/ClientStats.java:
```diff
@@ -0,0 +1,92 @@
+    }
+
+    @Override
+    public String toString() {
```
Went ahead and kept the old `toString`, but I'm wondering if it should be changed to also call `toString` on all the `HostStatistics` in the "statsPerHost" `Map`, or if that would be too verbose for a default.

Diagoras commented on client/src/main/java/org/asynchttpclient/netty/channel/ChannelManager.java:
```diff
@@ -488,4 +493,28 @@ public ChannelPool getChannelPool() {
         return eventLoopGroup;
     }
+
+    public ClientStats getClientStats() {
```
The implementation of this method does not make me happy, but I really don't see a better way to do this with Java 8 `Stream`s. If you think there's a better approach, let me know.

Diagoras commented on client/src/main/java/org/asynchttpclient/netty/channel/ChannelManager.java:
```diff
@@ -488,4 +493,28 @@ public ChannelPool getChannelPool() {
+                    .stream()
+                    .map(Channel::remoteAddress)
+                    .filter(a -> a.getClass() == InetSocketAddress.class)
+                    .map(a -> (InetSocketAddress) a)
```
So this sucks, but I didn't really see another way out. Are there any other types of socket that are likely to live here or in `openChannels`, or is this a mostly "safe" thing to do?

slandelle replied:
I think it's pretty safe to cast

Diagoras replied:
Yeah, the filter beforehand means we're not going to see an exception. I was just worried about leaving out other kinds of sockets from the stats, but it sounds like we're good.

Diagoras commented:
So, are we good to merge? And if so, should I squash my commits?

slandelle commented:
I hope to be able to review your latest changes in length this week.
As they break the API, they'll have to target AHC 2.1. I plan on switching master next week.
Thanks for your patience.
